src/tcache.c

```diff
@@ -169,6 +167,137 @@ tcache_gc_small(tsd_t *tsd, tcache_slow_t *tcache_slow, tcache_t *tcache,
+	}
+	assert(nremote_hard >= nremote && nremote_hard <= ncached);
+	/*
+	 * The number of ptrs outside of [addr_min, addr_max) is less than
+	 * the intended flush amount, shrink the range to [slab_min, slab_max).
+	 */
+	if (nremote < nflush) {
+		nremote = nremote_hard;
+		addr_min = slab_min;
+		addr_max = slab_max;
+	}
+	/*
+	 * Update the nflush to the larger value between the intended flush count
+	 * and the number of remote ptrs.
+	 */
+	nflush = nremote > nflush ? nremote : nflush;
+	assert(nflush < ncached || nremote == ncached);
+	if (nremote == 0 || nremote == ncached)	goto label_flush;
+
+	/*
+	 * Move the remote points to the bottom of the bin for flushing.
+	 * As long as moved to the bottom, the order of these nremote ptrs
+	 * does not matter, since they are going to be flushed anyway.
+	 * The rest of the ptrs are moved to the top of the bin, and their
+	 * relative order is not maintained.
+	 */
+	void **swap = NULL;
+	cache_bin_sz_t ntop = ncached - nremote, cnt = 0;
+	assert(ntop > 0 && ntop < ncached);
+	/*
+	 * Scan the [head, head + ntop) part of the cache bin, during which
+	 * bubbling the non-remote ptrs to the top of the bin.
+	 * After this, the [head, head + cnt) part of the bin contains only
+	 * non-remote ptrs, and they're in the same relative order as before.
+	 * While the [head + cnt, head + ntop) part contains only remote ptrs.
+	 */
+	for (void **cur = head; cur < head + ntop; cur++) {
+		assert(*cur != NULL);
+		bool not_remote = ((uintptr_t)*cur >= addr_min &&
+		    (uintptr_t)*cur < addr_max);
+		if (not_remote) {
+			/* Tracks the number of non-remote ptrs seen so far. */
+			cnt++;
+			/*
+			 * There is remote ptr before the current non-remote ptr,
+			 * swap the current non-remote ptr with the remote ptr,
+			 * and increment the swap pointer so that it's still
+			 * pointing to the top remote ptr in the bin.
+			 */
+			if (swap != NULL) {
+				assert(swap < cur);
+				assert((uintptr_t)*swap < addr_min ||
+				    (uintptr_t)*swap >= addr_max);
+				void *tmp = *cur;
+				*cur = *swap;
+				*swap = tmp;
+				swap++;
+				assert(swap <= cur);
+				assert((uintptr_t)*swap < addr_min ||
+				    (uintptr_t)*swap >= addr_max);
+			}
+			continue;
+		}
+
+		assert(!not_remote);
+		/* Swap always points to the top remote ptr in the bin. */
+		if (swap == NULL) {
+			swap = cur;
+		}
+	}
+	/*
+	 * Scan the [head + ntop, head + ncached) part of the cache bin,
+	 * after which it should only contain remote ptrs.
+	 */
+	for (void **cur = head + ntop; cur < head + ncached; cur++) {
+		assert(*cur != NULL);
+		/* Early break if all non-remote ptrs have been moved. */
+		if (cnt == ntop)	break;
+		if ((uintptr_t)*cur >= addr_min && (uintptr_t)*cur < addr_max) {
+			assert(*(head + cnt) != NULL && (
+			    (uintptr_t)*(head + cnt) < addr_min ||
+			    (uintptr_t)*(head + cnt) >= addr_max));
+			void *tmp = *cur;
+			*cur = *(head + cnt);
+			*(head + cnt) = tmp;
+			cnt++;
+		}
+	}
+	assert(cnt == ntop);
```

Comment:
There are non trivial scanning and reordering from the two loops above. We need another sanity checking loop that verifies the results (which should only include assertion in the body).

Reply:
Will the for loop be compiled away in opt build if only assertion in the loop body?